CGDisplayList-image-buffer-backed tiles grow without bound when repainted

hortont424 opened the pull request:
#### 9c49b078b670fd395a2848edee0478c6cd21f4ca
<pre>
CGDisplayList-image-buffer-backed tiles grow without bound when repainted
<a href="https://bugs.webkit.org/show_bug.cgi?id=243307">https://bugs.webkit.org/show_bug.cgi?id=243307</a>

Reviewed by Simon Fraser.

When repainting a display-list-backed tile, we shouldn&apos;t continue using the
existing recorder, but rather start a fresh one. Otherwise, the display list
just piles up.

Similarly, there is no reason to &quot;swap&quot; between display list image buffers,
since we repaint them fully every time, and this will cause trouble for future
resource caching efforts.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::clearContents):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::clearContents):
Add a &quot;clearContents&quot; ImageBuffer and ImageBufferBackend method.
I don&apos;t implement it for bitmap image buffers, opting instead for an assertion,
but you could *imagine* an implementation, so it seems OK to add here.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm:
(WebKit::GraphicsContextCGDisplayList::GraphicsContextCGDisplayList):
(WebKit::CGDisplayListImageBufferBackend::create):
Move platform context creation into the GraphicsContext constructor, so
that we can re-do both in one operation.

(WebKit::CGDisplayListImageBufferBackend::clearContents):
Implement clearContents as just swapping out the display list recorder.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::swapToValidFrontBuffer):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStore::Buffer::discard):
Move the display list buffer to its own toplevel, instead of hanging off Buffer,
since swapping display lists doesn&apos;t make any sense.

Canonical link: <a href="https://commits.webkit.org/252936@main">https://commits.webkit.org/252936@main</a>
</pre>

whsieh commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -478,6 +478,12 @@ void ImageBuffer::setVolatilityState(VolatilityState volatilityState)
 
+void ImageBuffer::clearContents()
+{
+    if (auto* backend = ensureBackendCreated())
+        backend->clearContents();
```
Is it intentional that this instantiates `m_backend` if it wasn't already created?

hortont424 replied:
Hmm, in this case it's true that that wouldn't make sense, but if you had a bitmap-backed surface and were implementing "clear contents" surely you'd want to make the backend in order to do so?

whsieh replied:
Ah, right — that makes sense!

hortont424 committed:
CGDisplayList-image-buffer-backed tiles grow without bound when repainted
https://bugs.webkit.org/show_bug.cgi?id=243307

Reviewed by Simon Fraser.

When repainting a display-list-backed tile, we shouldn't continue using the
existing recorder, but rather start a fresh one. Otherwise, the display list
just piles up.

Similarly, there is no reason to "swap" between display list image buffers,
since we repaint them fully every time, and this will cause trouble for future
resource caching efforts.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::clearContents):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::clearContents):
Add a "clearContents" ImageBuffer and ImageBufferBackend method.
I don't implement it for bitmap image buffers, opting instead for an assertion,
but you could *imagine* an implementation, so it seems OK to add here.

* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm:
(WebKit::GraphicsContextCGDisplayList::GraphicsContextCGDisplayList):
(WebKit::CGDisplayListImageBufferBackend::create):
Move platform context creation into the GraphicsContext constructor, so
that we can re-do both in one operation.

(WebKit::CGDisplayListImageBufferBackend::clearContents):
Implement clearContents as just swapping out the display list recorder.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::clearBackingStore):
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStore::swapToValidFrontBuffer):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
(WebKit::RemoteLayerBackingStore::paintContents):
(WebKit::RemoteLayerBackingStore::Buffer::discard):
Move the display list buffer to its own toplevel, instead of hanging off Buffer,
since swapping display lists doesn't make any sense.

Canonical link: https://commits.webkit.org/252936@main
